darts/models/forecasting/forecasting_model.py

```diff
@@ -834,34 +907,36 @@ def historical_forecasts(
 
             # iterate and forecast
             for _counter, pred_time in enumerate(iterator):
-                # build the training series
-                if min_timestamp > series_.time_index[0]:
+                # build the training series, as if retrain is True to break circular dependency
+                if min_timestamp_train > series_.time_index[0]:
                     train_series = series_.drop_before(
-                        min_timestamp - 1 * series_.freq
+                        min_timestamp_train - 1 * series_.freq
                     ).drop_after(pred_time)
                 else:
                     train_series = series_.drop_after(pred_time)
 
                 if train_length and len(train_series) > train_length:
                     train_series = train_series[-train_length:]
 
-                if retrain_func(
-                    counter=_counter,
-                    pred_time=pred_time,
-                    train_series=train_series,
-                    past_covariates=past_covariates_.drop_after(pred_time)
-                    if past_covariates_
-                    and ("past_covariates" in retrain_func_signature)
-                    else None,
-                    future_covariates=future_covariates_.drop_after(
-                        min(
-                            pred_time + train_series.freq * forecast_horizon,
-                            series_.end_time(),
+                # retrain_func processes the series that would be used for training
+                if (
+                    retrain_func(
+                        counter=_counter,
+                        pred_time=pred_time,
+                        train_series=train_series,
+                        past_covariates=past_covariates_.drop_after(pred_time)
+                        if past_covariates_
+                        else None,
+                        future_covariates=future_covariates_.drop_after(
+                            min(
+                                pred_time + train_series.freq * forecast_horizon,
+                                series_.end_time(),
+                            )
                         )
+                        if future_covariates_
+                        else None,
                     )
-                    if future_covariates_
-                    and ("future_covariates" in retrain_func_signature)
-                    else None,
+                    and pred_time in historical_forecasts_time_index_train
```

Comment:
`if` statements are executed from left to right. 
So we have to check like 

```
if pred_time in historical_forecasts_time_index_train and retrain_func(...):
```

An example: the following will raise an error because we cannot perform `>` on NoneTypes:
```
>>> x = None
>>> if x > 1 and x is not None:
...     pass
... else:
...     print(1)
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

But this one doesn't raise an error because we first check if `x` is not None:
```
>>> if x is not None and x > 1:
...     pass
... else:
...     print(1)
... 
1
...

Reply:
The two conditions were interchangeable and I picked this order because it was to cover the scenario where the `historical_forecasts_time_index_predict` starts before the `historical_forecasts_time_index_train` with a `retrain_func` returning `True` for the first `Timestamps`.

But if for some reason, the `retrain_func` returns unexpected values for this value of `pred_time`, then yes, your order would be more robust.

Reply:
The thing is we don't need to call retrain_func if the pred_time is not in the train indices. The way I suggest will not call retrain_func if it is not in the train index.

```
>>> pred_time = 0
>>> train_index = [1,2,3]
>>> def f():
...     print("called")
...     return True
... 
>>> if pred_time in train_index and f():
...     print("blabla")
... 
>>> if f() and pred_time in train_index:
...     print("blabla")
... 
called